llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -3024,3 +3024,170 @@ mccasformats::v1::loadDIETopLevel(DIETopLevelRef TopLevelRef) {
+  constexpr auto IsLittleEndian = true;
+  constexpr auto AddrSize = 8;
+  constexpr auto FormParams =
+      dwarf::FormParams{4 /*Version*/, AddrSize, dwarf::DwarfFormat::DWARF32};
+
+  while (true) {
+    Expected<dwarf::Attribute> Attr = AbbrevReader.readAttr();
+    if (!Attr)
+      return Attr.takeError();
+    if (*Attr == getEndOfAttributesMarker())
+      break;
+
+    Expected<dwarf::Form> Form = AbbrevReader.readForm();
+    if (!Form)
+      return Form.takeError();
+
+    Expected<uint64_t> FormSize =
+        getFormSize(*Form, FormParams, DIEData, DataReader.getOffset(),
+                    IsLittleEndian, AddrSize);
+    if (!FormSize)
+      return FormSize.takeError();
+
+    bool DataInDistinct = doesntDedup(*Form, *Attr);
+    auto &ReaderForData = DataInDistinct ? DistinctReader : DataReader;
+    ArrayRef<char> RawBytes;
+    if (auto E = ReaderForData.readArray(RawBytes, *FormSize))
+      return E;
+    AttrCallback(*Attr, *Form, toStringRef(RawBytes), DataInDistinct);
+  }
+  return Error::success();
+}
+
+static Expected<uint64_t> readAbbrevIdx(BinaryStreamReader &Reader) {
+  uint64_t Idx;
+  if (auto E = Reader.readULEB128(Idx))
+    return std::move(E);
+  return Idx;
+}
+
+static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+                                              unsigned AbbrevIdx) {
+  StringRef AbbrevData =
+      AbbrevEntries[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)];
+  return AbbrevEntryReader(AbbrevData);
+}
+
+Error DIEVisitor::visitDIERef(BinaryStreamReader &DataReader,
+                              unsigned AbbrevIdx, StringRef DIEData,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
+  AbbrevEntryReader AbbrevReader =
+      getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
+
+  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
+    StartTagCallback(*MaybeTag, AbbrevIdx);
+  else
+    return MaybeTag.takeError();
+
+  Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
+  if (!MaybeHasChildren)
+    return MaybeHasChildren.takeError();
+
+  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
+    return E;
+
+  if (!*MaybeHasChildren) {
+    EndTagCallback(false /*HadChildren*/);
+    return Error::success();
+  }
+
+  while (true) {
+    Expected<uint64_t> ChildAbbrevIdx = readAbbrevIdx(DistinctReader);
+    if (!ChildAbbrevIdx)
+      return ChildAbbrevIdx.takeError();
+
+    if (*ChildAbbrevIdx == getEndOfDIESiblingsMarker())
+      break;
+
+    if (*ChildAbbrevIdx == getDIEInAnotherBlockMarker()) {
+      if (auto E = visitDIERef(DIEChildrenStack.front()))
+        return E;
+      DIEChildrenStack = DIEChildrenStack.drop_front();
+      continue;
+    }
+
+    if (auto E =
+            visitDIERef(DataReader, *ChildAbbrevIdx, DIEData, DIEChildrenStack))
+      return E;
+  }
+
+  EndTagCallback(true /*HadChildren*/);
+  return Error::success();
+}
+
+Error DIEVisitor::visitDIERef(DIEDataRef Ref) {
```

Comment:
Can we rename this to something like `startDIERef`? Just curious if you think that makes more sense

Reply:
Good idea, will do!